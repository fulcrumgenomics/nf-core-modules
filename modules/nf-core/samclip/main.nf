process SAMCLIP {
    tag "${meta.id}"
    label 'process_low'

    conda "${moduleDir}/environment.yml"
    container "${workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container
        ? 'https://depot.galaxyproject.org/singularity/samclip:0.4.0--hdfd78af_1'
        : 'biocontainers/samclip:0.4.0--hdfd78af_1'}"

    input:
    tuple val(meta), path(sam)
    tuple val(meta2), path(reference), path(reference_index)

    output:
    tuple val(meta), path("*.sam"), emit: sam
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}.samclip"
    def is_compressed = reference.getName().endsWith(".gz")
    def ref_filename = reference.getName().replaceAll(/\.gz$/, "")
    """
    # decompress reference if gzipped
    ${is_compressed ? "gzip -c -d ${reference} > ${ref_filename}" : ""}

    samclip ${args} \\
        --ref ${ref_filename} \\
        < ${sam} \\
        > ${prefix}.sam \\

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        samclip: \$(echo \$(samclip --version 2>&1) | sed 's/^.*samclip //g' )
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}.samclip"
    """
    touch ${prefix}.sam

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        samclip: \$(echo \$(samclip --version 2>&1) | sed 's/^.*samclip //g' )
    END_VERSIONS
    """
}
