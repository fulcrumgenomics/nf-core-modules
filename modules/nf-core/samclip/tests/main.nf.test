nextflow_process {

    name "Test Process SAMCLIP"
    script "../main.nf"
    process "SAMCLIP"
    config "./nextflow.config"

    tag "modules"
    tag "modules_nfcore"
    tag "samclip"
    tag "samtools/view"

    setup {
        // Convert BAM to SAM file
        run("SAMTOOLS_VIEW") {
            script "../../samtools/view/main.nf"
            process {
                """
                // contains soft- and hard-clipped reads
                input[0] = [
                    [ id:'NA12878_chr22' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/NA12878.chr22.bam', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/NA12878.chr22.bam.bai', checkIfExists: true),

                ]
                input[1] = [[],[]]
                input[2] = []
                input[3] = []
                """
            }
        }
    }

    test("test-data - NA12878.chr22.bam") {
        config "./nextflow.config"

        when {
            process {
                """
                input[0] = SAMTOOLS_VIEW.out.sam

                input[1] = [
                    [ id:'chr22_ref' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true),
                ]

                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("test-data - NA12878.chr22.bam - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = SAMTOOLS_VIEW.out.sam

                input[1] = [
                    [ id:'chr22_ref' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true),
                ]

                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                        process.out,
                        process.out.versions.collect{ path(it).yaml }
                    ).match() }
            )
        }

    }

}