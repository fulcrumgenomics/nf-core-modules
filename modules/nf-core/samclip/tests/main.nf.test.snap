{
    "test-data - NA12878.chr22.bam": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "NA12878_chr22"
                        },
                        "NA12878_chr22.samclip.sam:md5,60a34f27ad09b38f457ade7ecc541692"
                    ]
                ],
                "1": [
                    "versions.yml:md5,535a0e68abc046e4aaefd32066af2119"
                ],
                "sam": [
                    [
                        {
                            "id": "NA12878_chr22"
                        },
                        "NA12878_chr22.samclip.sam:md5,60a34f27ad09b38f457ade7ecc541692"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,535a0e68abc046e4aaefd32066af2119"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-09-04T12:47:41.222016"
    },
    "test-data - NA12878.chr22.bam - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "NA12878_chr22"
                        },
                        "NA12878_chr22.samclip.sam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,535a0e68abc046e4aaefd32066af2119"
                ],
                "sam": [
                    [
                        {
                            "id": "NA12878_chr22"
                        },
                        "NA12878_chr22.samclip.sam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,535a0e68abc046e4aaefd32066af2119"
                ]
            },
            [
                {
                    "SAMCLIP": {
                        "samclip": "0.4.0"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-09-04T12:47:52.121328"
    }
}